File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/javascript/exercises/practice/connect/connect.js
export class Board {
  constructor(input) {
    this.board = input.map(row => row.trim().split(/\s+/));
    this.rows = this.board.length;
    this.cols = this.rows > 0 ? this.board[0].length : 0;
  }

  winner() {
    if (this.rows === 0 || this.cols === 0) return '';

    // Check O's top-to-bottom connection
    for (let c = 0; c < this.cols; c++) {
      if (this.board[0][c] === 'O' && this.bfs(0, c, 'O', (r, _) => r === this.rows - 1)) {
        return 'O';
      }
    }

    // Check X's left-to-right connection
    for (let r = 0; r < this.rows; r++) {
      if (this.board[r][0] === 'X' && this.bfs(r, 0, 'X', (_, c) => c === this.cols - 1)) {
        return 'X';
      }
    }

    return '';
  }

  bfs(startR, startC, player, isGoal) {
    const visited = Array.from({ length: this.rows }, () => 
      Array(this.cols).fill(false)
    );
    const queue = [[startR, startC]];
    visited[startR][startC] = true;

    const directions = [
      [-1, 0], [-1, 1],  // North, Northeast
      [0, -1], [0, 1],   // West, East
      [1, -1], [1, 0]    // Southwest, South
    ];

    while (queue.length > 0) {
      const [r, c] = queue.shift();
      // Check if we reached the correct edge for the player
      if (player === 'O' && r === this.rows - 1) return true;
      if (player === 'X' && c === this.cols - 1) return true;

      for (const [dr, dc] of directions) {
        const nr = r + dr;
        const nc = c + dc;
        if (nr >= 0 && nr < this.rows && nc >= 0 && nc < this.cols) {
          if (!visited[nr][nc] && this.board[nr][nc] === player) {
            visited[nr][nc] = true;
            queue.push([nr, nc]);
          }
        }
      }
    }
    return false;
  }
}

File: src/python-2025-02-27-deepclaude37-rust/python/exercises/practice/connect/connect.py

class ConnectGame:
    def __init__(self, board):
        # Parse the board into a 2D grid
        self.board = []
        for line in board.split('\n'):
            stripped = line.strip()
            if stripped:
                self.board.append(stripped.split())

    def get_winner(self):
        if not self.board:
            return ""
        
        rows = len(self.board)
        cols = len(self.board[0]) if rows > 0 else 0
        
        if rows == 0 or cols == 0:
            return ""

        # Check if player O has connected top to bottom
        visited = set()
        queue = []
        
        # Start from all O's in the top row
        for c in range(cols):
            if c < len(self.board[0]) and self.board[0][c] == 'O':
                queue.append((0, c))
                visited.add((0, c))
        
        # BFS to find path from top to bottom
        while queue:
            row, col = queue.pop(0)
            
            # Check if reached bottom row
            if row == rows - 1:
                return 'O'
            
            # Check all six possible hex grid neighbors
            for dr, dc in [(-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0)]:
                r, c = row + dr, col + dc
                if 0 <= r < rows and 0 <= c < len(self.board[r]):
                    if (r, c) not in visited and self.board[r][c] == 'O':
                        visited.add((r, c))
                        queue.append((r, c))

        # Check if player X has connected left to right
        visited = set()
        queue = []
        
        # Start from all X's in the leftmost column
        for r in range(rows):
            if self.board[r][0] == 'X':
                queue.append((r, 0))
                visited.add((r, 0))
        
        # BFS to find path from left to right
        while queue:
            row, col = queue.pop(0)
            
            # Check if reached rightmost column
            if col == len(self.board[row]) - 1:
                return 'X'
            
            # Check all six possible hex grid neighbors
            for dr, dc in [(-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0)]:
                r, c = row + dr, col + dc
                if 0 <= r < rows and 0 <= c < len(self.board[r]):
                    if (r, c) not in visited and self.board[r][c] == 'X':
                        visited.add((r, c))
                        queue.append((r, c))

        # No winner found
        return ""
